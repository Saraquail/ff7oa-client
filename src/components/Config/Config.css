:root {
  --URC-Red: 0;
  --URC-Green: 0;
  --URC-Blue: 0;
  --ULC-Red: 0;
  --ULC-Green: 0;
  --ULC-Blue: 0;
  --LRC-Red: 0;
  --LRC-Green: 0;
  --LRC-Blue: 0;
  --LLC-Red: 0;
  --LLC-Green: 0;
  --LLC-Blue: 0;
}

.config-menu {
  --opacity: 0.5;

  --URC: rgba(
    var(--URC-Red),
    var(--URC-Green),
    var(--URC-Blue),
    var(--opacity)
  );
  --ULC: rgba(
    var(--ULC-Red),
    var(--ULC-Green),
    var(--ULC-Blue),
    var(--opacity)
  );
  --LRC: rgba(
    var(--LRC-Red),
    var(--LRC-Green),
    var(--LRC-Blue),
    var(--opacity)
  );
  --LLC: rgba(
    var(--LLC-Red),
    var(--LLC-Green),
    var(--LLC-Blue),
    var(--opacity)
  );

  background-color: black;
  background: 
    radial-gradient(ellipse at top left, var(--ULC), transparent),
    radial-gradient(ellipse at bottom right, var(--LRC), transparent),
    radial-gradient(ellipse at top right, var(--URC), transparent),
    radial-gradient(ellipse at bottom left, var(--LLC), transparent),
    black;

}