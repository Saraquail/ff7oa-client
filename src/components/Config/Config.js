import React, { useState } from 'react';
import Nav from '../Nav/Nav';
import './Config.css';
// this component is for a future goal of user configuring
// theme settings based on the menu config in the game
// it works as is, and is a fun nostaligic experience for
// people who are familiar with Final Fantasy 7.

const INITIAL_STATE = {
  URC_open: false,
  ULC_open: false,
  LRC_open: false,
  LLC_open: false,
  URC: {
    Red: '0',
    Green: '0',
    Blue: '128',
  },
  ULC: {
    Red: '0',
    Green: '0',
    Blue: '173',
  },
  LRC: {
    Red: '0',
    Green: '0',
    Blue: '32',
  },
  LLC: {
    Red: '0',
    Green: '0',
    Blue: '88',
  },
};

const Config = () => {
  const [state, setState] = useState(INITIAL_STATE);

  /* currying */
  const handleColorChange = (corner, color) => (e) => {
    const values = state[corner];
    const newValues = { ...values, [color]: e.target.value };
    setState({ ...state, [corner]: { ...newValues } });
  };

  const handleClick = (ev) => {
    const toggleCorner = ev.target.value;
    const toggleKey = `${toggleCorner}_open`;
    setState((prevState) => ({
      ...state,
      [toggleKey]: !prevState[toggleKey],
    }));
  };

  const corners = ['URC', 'ULC', 'LRC', 'LLC'];
  const colors = ['Red', 'Green', 'Blue'];
  const labels = {
    URC: 'Upper Right Corner',
    ULC: 'Upper Left Corner',
    LRC: 'Lower Right Corner',
    LLC: 'Lower Left Corner',
  };

  const input = corners.map((corner) => colors.map((color) => (
    <div
      className={`${corner}`}
      key={`${corner}${color}`}
    >
      <label htmlFor={`${corner}${color}`}>
        {`${labels[corner]} ${color} value`}
      </label>
      <input
        className={`${corner}`}
        name={`${corner}${color}`}
        id={`${corner}${color}`}
        type="range"
        min="0"
        max="255"
        onChange={handleColorChange(corner, color)}
      />
      {state[corner][color]}
    </div>
  )));

  // creates CSS custom prop for each corner and color
  // then creates an array - 1st value is property name, second value is the corner and colors
  // @example
  /*
    [
      ['--URC-Red', 0],
      ['--URC-Green', 22],
      ...
    ]
  */
  const styleArr = corners.map((corner) => colors.map((color) => [`--${corner}-${color}`, state[corner][color]]));

  // takes an array of [key, value] and converts it to an object of
  // key/value pairs like { key: value }
  const styles = styleArr
    .reduce((obj, [red, green, blue]) => ({
      ...obj,
      [red[0]]: red[1],
      [green[0]]: green[1],
      [blue[0]]: blue[1],
    }), {});

  return (
    <div className="config-menu" style={styles}>
      <Nav />
      <p className="onboarding">
        This is an experimental feature.
        Play around as you like and adjust the background
        color for this page just like in the original game!
        Have fun, and come back soon to see this implemented around the site.
      </p>
      <button type="button" className="input-group" value="URC" onClick={handleClick}>
        {state.URC_open ? 'Close Upper Right Corner' : 'Adjust Upper Right Corner' }
      </button>
      {state.URC_open ? input[0] : '' }

      <button type="button" className="input-group" value="ULC" onClick={handleClick}>
        {state.ULC_open ? 'Close Upper Left Corner' : 'Adjust Upper Left Corner' }
      </button>
      {state.ULC_open ? input[1] : ''}

      <button type="button" className="input-group" value="LRC" onClick={handleClick}>
        {state.LRC_open ? 'Close Lower Right Corner' : 'Adjust Lower Right Corner' }
      </button>
      {state.LRC_open ? input[2] : ''}

      <button type="button" className="input-group" value="LLC" onClick={handleClick}>
        {state.LLC_open ? 'Close Lower Left Corner' : 'Adjust Lower Left Corner' }
      </button>
      {state.LLC_open ? input[3] : ''}
    </div>
  );
};

export default Config;
