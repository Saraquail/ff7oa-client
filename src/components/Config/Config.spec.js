import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom'
import renderer from 'react-test-renderer';

import Config from './Config.js'

describe('<Config/>', () => {

  it('renders without crashing', () => {
  const div = document.createElement('div')

  ReactDOM.render(
  <BrowserRouter>
    <Config />
  </BrowserRouter>, div)

  ReactDOM.unmountComponentAtNode(div)
  })

  it('renders the UI as expected', () => {
    const tree = renderer
  .create(
    <BrowserRouter>
      <Config/>
    </BrowserRouter>
  )
  .toJSON();
    expect(tree).toMatchSnapshot();  
  })
})
