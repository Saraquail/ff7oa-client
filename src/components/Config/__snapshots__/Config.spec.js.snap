// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Config/> renders the UI as expected 1`] = `
<div
  className="config-menu"
  style={
    Object {
      "--LLC-Blue": "88",
      "--LLC-Green": "0",
      "--LLC-Red": "0",
      "--LRC-Blue": "32",
      "--LRC-Green": "0",
      "--LRC-Red": "0",
      "--ULC-Blue": "173",
      "--ULC-Green": "0",
      "--ULC-Red": "0",
      "--URC-Blue": "128",
      "--URC-Green": "0",
      "--URC-Red": "0",
    }
  }
>
  <nav>
    <div
      className="header-image-container"
    >
      <img
        alt="green orb of magic materia from final fantasy 7"
        className="materia-img"
        src="Magic_Materia.png"
      />
      <img
        alt="purple orb of complete materia from final fantasy 7"
        className="materia-img"
        src="Complete_Materia.png"
      />
      <img
        alt="yellow orb of command materia from final fantasy 7"
        className="materia-img"
        src="Command_Materia.png"
      />
      <img
        alt="blue orb of support materia from final fantasy 7"
        className="materia-img"
        src="Support_Materia.png"
      />
      <img
        alt="red orb of summoning materia from final fantasy 7"
        className="materia-img"
        src="Summoning_Materia.png"
      />
    </div>
    <ul>
      <li>
        <a
          href="/bestiary"
          onClick={[Function]}
        >
          Bestiary
        </a>
      </li>
      <li>
        <a
          href="/PHS"
          onClick={[Function]}
        >
          My PHS
        </a>
      </li>
      <li>
        <a
          href="/Config"
          onClick={[Function]}
        >
          Config
        </a>
      </li>
      <li>
        <a
          href="/"
          onClick={[Function]}
        >
          Login
        </a>
      </li>
    </ul>
  </nav>
  <p
    className="onboarding"
  >
    This is an experimental feature. Play around as you like and adjust the background color for this page just like in the original game! Have fun, and come back soon to see this implemented around the site.
  </p>
  <button
    className="input-group"
    onClick={[Function]}
    type="button"
    value="URC"
  >
    Adjust Upper Right Corner
  </button>
  
  <button
    className="input-group"
    onClick={[Function]}
    type="button"
    value="ULC"
  >
    Adjust Upper Left Corner
  </button>
  
  <button
    className="input-group"
    onClick={[Function]}
    type="button"
    value="LRC"
  >
    Adjust Lower Right Corner
  </button>
  
  <button
    className="input-group"
    onClick={[Function]}
    type="button"
    value="LLC"
  >
    Adjust Lower Left Corner
  </button>
  
</div>
`;
